src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -1481,7 +1481,14 @@ private BoundExpression BindDefaultExpression(DefaultExpressionSyntax node, Diag
                 expression = null;
                 if (node is IdentifierNameSyntax identifier)
                 {
-                    var type = BindNativeIntegerSymbolIfAny(identifier, diagnostics);
+                    // Don't bind nameof(nint) or nameof(nuint) so that ERR_NameNotInContext is reported.
+                    // The check for ArgumentSyntax is to allow the binding for something like nameof(nint.Equals).
+                    var type = IsInsideNameof && identifier.Parent is ArgumentSyntax &&
+                        identifier.Parent?.Parent?.Parent is InvocationExpressionSyntax invocation &&
+                        (invocation.Expression as IdentifierNameSyntax)?.Identifier.ContextualKind() == SyntaxKind.NameOfKeyword
+                        ? null
+                        : BindNativeIntegerSymbolIfAny(identifier, diagnostics);
```

Comment:
nit: could we move all the extra logic into `BindNativeIntegerSymbolIfAny`? It would help encapsulate the complexity away from sight a bit.

Reply:
@jcouv There is another call for `BindNativeIntegerSymbolIfAny`:

https://github.com/dotnet/roslyn/blob/45f2aedcfe54c9b0cc6d9255e3b8945942359f1b/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs#L893

I don't know if there would be unintended effects of changing the method itself. That's why I kept the change outside it to only affect this code path.

If you can confirm that both method calls need this change, I'll do that.

Reply:
The second caller doesn't need this change, but won't be hurt (the `IsInsideNameof` check is fast).

@cston Just to confirm, what do you think of my proposal (move the new logic inside `BindNativeIntegerSymbolIfAny`)?

Reply:
> what do you think of my proposal (move the new logic inside `BindNativeIntegerSymbolIfAny`)?

Sounds reasonable. And it would mean we could avoid the extra check unless the identifier is "nint" or "nuint".

---
In reply to: [485950715](https://github.com/dotnet/roslyn/pull/47497#discussion_r485950715) [](ancestors = 485950715)